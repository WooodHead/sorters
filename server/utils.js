function prepare(o) {
    if (o && o._id) {
        o._id = o._id.toString()
    }
    return o
}

async function findOne(Collection, ...args) {
    return prepare(await Collection.findOne(...args))
}

async function find(Collection, ...args) {
    return (await Collection.find(...args).toArray()).map(prepare)
}

module.exports = {
    prepare,
    findOne,
    find,
}