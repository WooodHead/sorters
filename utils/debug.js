
export function log(arg) {
    console.log(arg)
    return arg
}
