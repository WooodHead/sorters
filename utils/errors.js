export const errorMessage = (e) => {
    if (e.graphQLErrors && e.graphQLErrors.length > 0) {
        return e.graphQLErrors[0].message
    } else {
        return e.message
    }
}