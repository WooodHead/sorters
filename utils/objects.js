
function toObject(arr) {
    const res = {}
    for (const [key, value] of arr) {
        res[key] = value
    }
    return res
}

function objectMap(object, keyMap = key => key, valueMap = value => value) {
    return Object.keys(object).reduce((o, key) => {
        o[keyMap(key)] = valueMap(object[key])
        return o
    }, {})
}

module.exports = {
    toObject,
    objectMap
}
