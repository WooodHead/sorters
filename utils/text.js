export const replaceAll = (string, search, replacement) => {
  if (typeof string === "string") {
    return string.replace(new RegExp(search, 'g'), replacement)
  } else {
    return string
  }
}

export const teaserAndInfo = (content, min, opt, max) => {
	if (content) {
		if (content.length <= max) {
			return {
				cut: false,
				teaser: content
			};
		}
		const words = content.split(" ");
		let teaser = "";
		for (let i=0; i<words.length; i++) {
			const candidate = words[i];
			let wouldbelength = teaser.length + 1 + candidate.length + 3;
			if (wouldbelength >= opt) {
				// We're done
				if (wouldbelength > max) {
					if (teaser.length + 3 < min) {
						// cut the word
						return {
							teaser: teaser + " " + candidate.substring(0, opt - teaser.length - 1) + "...",
							cut: true,
						};
					} else {
						return {
							teaser: teaser + "...",
							cut: true,
						};
					}
				} else {
					return {
						teaser: teaser + " " + candidate + "...",
						cut: true,
					};
				}
			}
			teaser += " " + candidate;
		}
		return {
			teaser: teaser + "...",
			cut: true,
		};
	} else {
		return {
			teaser: "",
			cut: false,
		};
	}
};
export const teaser = (content, min, opt, max) => {
	const teaser = teaserAndInfo(content, min, opt, max);
	return teaser.teaser;
}
