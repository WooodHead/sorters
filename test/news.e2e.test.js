import cheerio from 'cheerio'
import {pretty} from './utils'
import {ObjectID} from 'mongodb'
import {setup, teardown} from './setup'
import {generateAndLogUser, setUserData, getGoalIds} from './fixtures'
import {request} from './utils'

let db
let browser

beforeAll(async () => {
    const res = await setup()
    db = res.db
    browser = res.browser
})


afterAll(async () => {
    await teardown({db, browser})
})

describe('news', () => {
    afterEach(async () => {
        await db.dropDatabase()    
    })

    it('displays', async () => {
        const browserPage = await browser.createPage()
        const status = await browserPage.open(`http://localhost:3000/news`)
        expect(status).toBe('success')

        const text = await browserPage.property('content')
        const $ = cheerio.load(text)
        const page = $('#__next')
        const html = pretty(page.html())
        expect(html).toMatchSnapshot()
    })

    it('displays read events', async () => {
        const Users = db.collection('users')
        const Reads = db.collection('reads')
        const Goals = db.collection('goals')
        const {insertedId: userId} = await Users.insertOne({
            local: {
                username: 'test'
            },
        });
        const {insertedId: goalId} = await Goals.insertOne({
            userId,
            title: 'Test',
        })
        const {insertedId: readId} = await Reads.insertOne({
            userId,
            title: 'TestGoal',
        })
        await Users.update({}, {
            $set: {
                readIds: [readId],
                goalIds: [goalId],
            },
        })
        const Events = db.collection('events')
        const date = new Date('2017-08-24T06:52:59.645Z')
        await Events.insertOne({
            userId,
            type: 'created-read',
            date,
            readId,
        })
        await Events.insertOne({
            userId,
            type: 'reading-read',
            date,
            readId,
        })
        await Events.insertOne({
            userId,
            type: 'read-read',
            date,
            readId,
        })
        await Events.insertOne({
            userId,
            type: 'created-goal',
            date,
            goalId,
        })
        await Events.insertOne({
            userId,
            type: 'doing-goal',
            date,
            goalId,
        })
        await Events.insertOne({
            userId,
            type: 'done-goal',
            date,
            goalId,
        })
        await Events.insertOne({
            userId,
            type: 'done-goal',
            date,
            goalId,
        })
        const browserPage = await browser.createPage()
        const status = await browserPage.open(`http://localhost:3000/news`)
        expect(status).toBe('success')
        
        const text = await browserPage.property('content')
        const $ = cheerio.load(text)
        const page = $('#__next')
        const html = pretty(page.html())
        expect(html).toMatchSnapshot()
    })

    it('displays journal events', async () => {
        const browserPage = await browser.createPage()

        await generateAndLogUser(browserPage)

        const goalIds = await getGoalIds(browserPage)

        await request(browserPage, `http://localhost:3000/graphql`, 'POST', {
            query: `
                mutation($entry: NewEntryInput!) {
                    createEntry(entry: $entry) {
                        _id
                    }
                }
            `,
            variables: {
                entry: {
                    title: 'Hello World',
                    url: 'http://www.example.com',
                    description: 'Lorem ipsum',
                    goalIds,
                },
            },
        })

        const status = await browserPage.open(`http://localhost:3000/news`)
        expect(status).toBe('success')

        const text = await browserPage.property('content')
        const $ = cheerio.load(text)
        const page = $('#__next')
        const html = pretty(page.html())
        expect(html).toMatchSnapshot()
    })

    it('displays changes to user data', async () => {
        const browserPage = await browser.createPage()

        await generateAndLogUser(browserPage)
        await setUserData(browserPage)

        const status = await browserPage.open(`http://localhost:3000/news`)
        expect(status).toBe('success')

        const text = await browserPage.property('content')
        const $ = cheerio.load(text)
        const page = $('#__next')
        const html = pretty(page.html())
        expect(html).toMatchSnapshot()
    })

    it('limits results', async () => {
        const browserPage = await browser.createPage()
        
        await generateAndLogUser(browserPage)
        const goalIds = await getGoalIds(browserPage)

        for (let i=0; i<11; i++) {
            await request(browserPage, `http://localhost:3000/graphql`, 'POST', {
                query: `
                    mutation($entry: NewEntryInput!) {
                        createEntry(entry: $entry) {
                            _id
                        }
                    }
                `,
                variables: {
                    entry: {
                        title: 'Hello World',
                        url: 'http://www.example.com',
                        description: 'Lorem ipsum',
                        goalIds,
                    },
                },
            })    
        }

        const status = await browserPage.open(`http://localhost:3000/news`)
        expect(status).toBe('success')

        const text = await browserPage.property('content')
        const $ = cheerio.load(text)
        const page = $('#__next')
        const html = pretty(page.html())
        expect(html).toMatchSnapshot()                
    })
})

