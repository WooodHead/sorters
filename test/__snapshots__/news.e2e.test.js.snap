// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`news displays 1`] = `
"<div data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"-1029702313\\">
  <div data-reactid=\\"2\\">
    <!-- react-empty: 3 -->
    <div style=\\"padding-top:50px;padding-bottom:20px;\\" data-reactid=\\"4\\">
      <nav class=\\"navbar navbar-inverse navbar-fixed-top\\" data-reactid=\\"5\\">
        <div class=\\"container\\" data-reactid=\\"6\\">
          <div class=\\"navbar-header\\" data-reactid=\\"7\\">
            <button type=\\"button\\" class=\\"navbar-toggle collapsed\\" data-toggle=\\"collapse\\" data-target=\\"#navbar\\" aria-expanded=\\"false\\" aria-controls=\\"navbar\\" data-reactid=\\"8\\"><span class=\\"sr-only\\" data-reactid=\\"9\\">Toggle navigation</span><span class=\\"icon-bar\\" data-reactid=\\"10\\"></span><span class=\\"icon-bar\\" data-reactid=\\"11\\"></span><span class=\\"icon-bar\\" data-reactid=\\"12\\"></span></button>
            <a class=\\"navbar-brand\\"
              href=\\"/\\" data-reactid=\\"13\\">Sorters Club</a>
          </div>
          <div id=\\"navbar\\" class=\\"collapse navbar-collapse\\" data-reactid=\\"14\\">
            <ul class=\\"nav navbar-nav navbar-right\\" data-reactid=\\"15\\">
              <li class=\\"false\\" data-reactid=\\"16\\">
                <a href=\\"/\\" data-reactid=\\"17\\">Home</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"18\\">
                <a href=\\"/about\\" data-reactid=\\"19\\">About</a>
              </li>
              <li class=\\"active\\" data-reactid=\\"20\\">
                <a href=\\"/news\\" data-reactid=\\"21\\">News</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"22\\">
                <a href=\\"/users\\" data-reactid=\\"23\\">Sorters</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"24\\">
                <a href=\\"/dashboard\\" data-reactid=\\"25\\">Dashboard</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"26\\">
                <a href=\\"/login\\" data-reactid=\\"27\\">Log in</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"28\\">
                <a href=\\"/register\\" data-reactid=\\"29\\">Register</a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </div>
    <div class=\\"container\\" data-reactid=\\"30\\">
      <h1 data-reactid=\\"31\\">News</h1>
      <p data-reactid=\\"32\\">See what happened recently in the community.</p>
      <div data-reactid=\\"33\\"></div>
    </div>
    <footer data-reactid=\\"34\\">
      <div class=\\"container\\" data-reactid=\\"35\\">
        <hr data-reactid=\\"36\\">
        <p data-reactid=\\"37\\">
          <!-- react-text: 38 -->Brought to you with &#x2764; by
          <!-- /react-text -->
          <a href=\\"/about\\" data-reactid=\\"39\\">Nick Redmark</a>
          <!-- react-text: 40 -->. Support and discuss the development of this platform on
          <!-- /react-text -->
          <a href=\\"https://www.patreon.com/nickredmark\\" target=\\"_blank\\" data-reactid=\\"41\\">Patreon</a>
          <!-- react-text: 42 -->.
          <!-- /react-text -->
        </p>
      </div>
    </footer>
  </div>
</div>"
`;

exports[`news displays read events 1`] = `
"<div data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"901455493\\">
  <div data-reactid=\\"2\\">
    <!-- react-empty: 3 -->
    <div style=\\"padding-top:50px;padding-bottom:20px;\\" data-reactid=\\"4\\">
      <nav class=\\"navbar navbar-inverse navbar-fixed-top\\" data-reactid=\\"5\\">
        <div class=\\"container\\" data-reactid=\\"6\\">
          <div class=\\"navbar-header\\" data-reactid=\\"7\\">
            <button type=\\"button\\" class=\\"navbar-toggle collapsed\\" data-toggle=\\"collapse\\" data-target=\\"#navbar\\" aria-expanded=\\"false\\" aria-controls=\\"navbar\\" data-reactid=\\"8\\"><span class=\\"sr-only\\" data-reactid=\\"9\\">Toggle navigation</span><span class=\\"icon-bar\\" data-reactid=\\"10\\"></span><span class=\\"icon-bar\\" data-reactid=\\"11\\"></span><span class=\\"icon-bar\\" data-reactid=\\"12\\"></span></button>
            <a class=\\"navbar-brand\\"
              href=\\"/\\" data-reactid=\\"13\\">Sorters Club</a>
          </div>
          <div id=\\"navbar\\" class=\\"collapse navbar-collapse\\" data-reactid=\\"14\\">
            <ul class=\\"nav navbar-nav navbar-right\\" data-reactid=\\"15\\">
              <li class=\\"false\\" data-reactid=\\"16\\">
                <a href=\\"/\\" data-reactid=\\"17\\">Home</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"18\\">
                <a href=\\"/about\\" data-reactid=\\"19\\">About</a>
              </li>
              <li class=\\"active\\" data-reactid=\\"20\\">
                <a href=\\"/news\\" data-reactid=\\"21\\">News</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"22\\">
                <a href=\\"/users\\" data-reactid=\\"23\\">Sorters</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"24\\">
                <a href=\\"/dashboard\\" data-reactid=\\"25\\">Dashboard</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"26\\">
                <a href=\\"/login\\" data-reactid=\\"27\\">Log in</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"28\\">
                <a href=\\"/register\\" data-reactid=\\"29\\">Register</a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </div>
    <div class=\\"container\\" data-reactid=\\"30\\">
      <h1 data-reactid=\\"31\\">News</h1>
      <p data-reactid=\\"32\\">See what happened recently in the community.</p>
      <div data-reactid=\\"33\\">
        <div data-reactid=\\"34\\">
          <div style=\\"margin-bottom:24px;display:flex;\\" data-reactid=\\"35\\">
            <div data-reactid=\\"36\\">
              <a href=\\"/u/test\\" data-reactid=\\"37\\">
                <img alt=\\"Gravatar for undefined\\" style=\\"margin-right:24px;\\" src=\\"//www.gravatar.com/avatar/test?d=retro&amp;r=g&amp;s=75\\" srcset=\\"//www.gravatar.com/avatar/test?d=retro&amp;r=g&amp;s=150 2x\\" height=\\"75\\" width=\\"75\\" class=\\"react-gravatar\\"
                  data-reactid=\\"38\\">
              </a>
            </div>
            <div data-reactid=\\"39\\">
              <h4 style=\\"margin-top:0;\\" data-reactid=\\"40\\">
                <a href=\\"/u/test\\" data-reactid=\\"41\\">test</a>
              </h4>
              <ul style=\\"padding-left:0;list-style-position:inside;\\" data-reactid=\\"42\\">
                <li data-reactid=\\"43\\">
                  <!-- react-text: 44 -->Achieved
                  <!-- /react-text --><span data-reactid=\\"45\\"><!-- react-text: 46 --><!-- /react-text --><!-- react-text: 47 -->&#xA0;<!-- /react-text --><em data-reactid=\\"48\\">TestGoal</em></span></li>
                <li data-reactid=\\"49\\">
                  <!-- react-text: 50 -->Spoke about
                  <!-- /react-text --><span data-reactid=\\"51\\"><!-- react-text: 52 --><!-- /react-text --><!-- react-text: 53 -->&#xA0;<!-- /react-text --><a href=\\"http://example.com/video\\" data-reactid=\\"54\\">Test</a></span></li>
                <li data-reactid=\\"55\\">
                  <!-- react-text: 56 -->Wrote about
                  <!-- /react-text --><span data-reactid=\\"57\\"><!-- react-text: 58 --><!-- /react-text --><!-- react-text: 59 -->&#xA0;<!-- /react-text --><a href=\\"http://example.com/article\\" data-reactid=\\"60\\">Test</a></span></li>
                <li data-reactid=\\"61\\">
                  <!-- react-text: 62 -->Read
                  <!-- /react-text --><span data-reactid=\\"63\\"><!-- react-text: 64 --><!-- /react-text --><!-- react-text: 65 -->&#xA0;<!-- /react-text --><em data-reactid=\\"66\\">Test</em></span></li>
                <li data-reactid=\\"67\\">
                  <!-- react-text: 68 -->Is working on
                  <!-- /react-text --><span data-reactid=\\"69\\"><!-- react-text: 70 --><!-- /react-text --><!-- react-text: 71 -->&#xA0;<!-- /react-text --><em data-reactid=\\"72\\">TestGoal</em></span></li>
                <li data-reactid=\\"73\\">
                  <!-- react-text: 74 -->Started reading
                  <!-- /react-text --><span data-reactid=\\"75\\"><!-- react-text: 76 --><!-- /react-text --><!-- react-text: 77 -->&#xA0;<!-- /react-text --><em data-reactid=\\"78\\">Test</em></span></li>
                <li data-reactid=\\"79\\">
                  <!-- react-text: 80 -->Wants to achieve
                  <!-- /react-text --><span data-reactid=\\"81\\"><!-- react-text: 82 --><!-- /react-text --><!-- react-text: 83 -->&#xA0;<!-- /react-text --><em data-reactid=\\"84\\">TestGoal</em></span></li>
                <li data-reactid=\\"85\\">
                  <!-- react-text: 86 -->Added books to reading list:
                  <!-- /react-text --><span data-reactid=\\"87\\"><!-- react-text: 88 --><!-- /react-text --><!-- react-text: 89 -->&#xA0;<!-- /react-text --><em data-reactid=\\"90\\">Test</em></span></li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
    <footer data-reactid=\\"91\\">
      <div class=\\"container\\" data-reactid=\\"92\\">
        <hr data-reactid=\\"93\\">
        <p data-reactid=\\"94\\">
          <!-- react-text: 95 -->Brought to you with &#x2764; by
          <!-- /react-text -->
          <a href=\\"/about\\" data-reactid=\\"96\\">Nick Redmark</a>
          <!-- react-text: 97 -->. Support and discuss the development of this platform on
          <!-- /react-text -->
          <a href=\\"https://www.patreon.com/nickredmark\\" target=\\"_blank\\" data-reactid=\\"98\\">Patreon</a>
          <!-- react-text: 99 -->.
          <!-- /react-text -->
        </p>
      </div>
    </footer>
  </div>
</div>"
`;
